src/gpu-tips-n-tricks.md

```diff
@@ -1,61 +1,293 @@
+
+`error C2131: expression did not evaluate to a constant`
+
+A simple fix for this error is to declare the `constexpr` variable as `static constexpr`.
+```c++
+static constexpr P = ...; // Omitting the static would result in an error on MSVC
+MFEM_FORALL(I,N,
+{
+  double my_data[P];
+});
+```
+Similar problems and workarounds are discussed [here](https://stackoverflow.com/questions/55136414).
+
+## Error: `alias not found`:
+This error message indicates that you are trying to move an "alias" `Vector` to gpu while its "base" `Vector` did not have gpu allocation (valid or not) when the alias was created (and may still not have gpu allocation when the move of the "alias" was attempted). This is another case where we cannot update the "base" `Vector` because we do not have access to it, and even if we did, there are other complications.
+
+Therefore, we need to follow this rule:
+**if you are creating an "alias" that will be used on device, you need to ensure that the "base" is allocated on device.**
+
+Depending on the context, one can use different methods to do that.
+For example, if the "base" is initialized (on host, otherwise there will be no issue) in the same function that will create the alias, one can call `base.Write()` to create the device allocation followed by `base.HostWrite()` and then initialize "base" on host -- this sequence avoids any unnecessary host-device transfers.
+
+Another example: if the "base" was initialized outside of the function where the "alias" is created, then the most appropriate choice probably is to call `base.Read()` before creating the "alias". Since the alias will need the data on device, the incurred host-to-device transfer is (at least partially) necessary anyway.
+
+Ideally, "base" Vectors that will be modified/accessed on device through aliases should be allocated on device to begin with, e.g. using `Vector::SetSize(int s, MemoryType mt)` typically with `mt = Device::GetDeviceMemoryType()`.
+
+## `MakeRef()` vectors do not see the same valid host/device data as their base vector:
+```c++
+const int vSize = 10;
+Vector v; v.UseDevice(true); v.SetSize(vSize); v = 0.0;
+cout << "IsHost(v) = " << IsHostMemory(v.GetMemory().GetMemoryType()) << endl;
+
+Vector w; w.MakeRef(v, 0, vSize);
+cout << "IsHost(w) = " << IsHostMemory(w.GetMemory().GetMemoryType()) << endl;
+
+auto hv = v.HostWrite();
+for (int j = 0; j < vSize; j++) { hv[j] = 1.0; }
+cout << "IsHost(v) = " << IsHostMemory(v.GetMemory().GetMemoryType()) << endl;
+cout << "IsHost(w) = " << IsHostMemory(w.GetMemory().GetMemoryType()) << endl;
+
+Vector z; z.UseDevice(true); z.SetSize(vSize);
+auto dz = z.Write();
+auto dw = w.Read();
+MFEM_FORALL(i, vSize,
+{
+  dz[i] = dw[i];
+});
+z.HostRead();
+cout << "norm(z) = " << z.Norml2() << endl;
+
+dz = z.Write();
+auto dv = v.Read();
+MFEM_FORALL(i, vSize,
+{
+  dz[i] = dv[i];
+});
+z.HostRead();
+cout << "norm(z) = " << z.Norml2() << endl;
+```
```

Comment:
The output of the above example would add a lot of value to it, I think.

Reply:
I added the output, though I'm a bit confused by this example, @v-dobrev can you check that the output is what you expected, should we be observing some difference in the output?

Reply:
Did you get this output with `"cuda"` device configuration? I'd expect different output in that case.